File: src/foam/modules/icloud-sync/ios/ICloudSyncModule.swift
import ExpoModulesCore
import Foundation

public class ICloudSyncModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ICloudSync")

    Function("isAvailable") {
      FileManager.default.ubiquityIdentityToken != nil
    }

    AsyncFunction("getString") { (key: String) -> String? in
      NSUbiquitousKeyValueStore.default.synchronize()
      return NSUbiquitousKeyValueStore.default.string(forKey: key)
    }

    AsyncFunction("setString") { (key: String, value: String) in
      let store = NSUbiquitousKeyValueStore.default
      store.set(value, forKey: key)
      store.synchronize()
    }

    AsyncFunction("remove") { (key: String) in
      let store = NSUbiquitousKeyValueStore.default
      store.removeObject(forKey: key)
      store.synchronize()
    }

    AsyncFunction("synchronize") {
      NSUbiquitousKeyValueStore.default.synchronize()
    }
  }
}

File: modules/icloud-sync/android/src/main/java/expo/modules/icloudsync/ICloudSyncModule.kt
package expo.modules.icloudsync

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ICloudSyncModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ICloudSync")

    Function("isAvailable") {
      false
    }

    AsyncFunction("getString") { _: String ->
      null as String?
    }

    AsyncFunction("setString") { _: String, _: String ->
    }

    AsyncFunction("remove") { _: String ->
    }

    AsyncFunction("synchronize") {
      false
    }
  }
}
